core/src/main/scala/com/avsystem/commons/misc/Opt.scala

```diff
@@ -14,6 +14,8 @@ object Opt {
     if (value != null) new Opt[A](value)
     else throw new NullPointerException
 
+  def when[A](cond: Boolean)(value: => Opt[A]): Opt[A] = if (cond) value else Opt.Empty
```

Comment:
I know we're spotty with docs but you can be better. Let's reuse `scala.Option#when` docs (I'm actually suprised this was not your inspiration :D)

Reply:
Right, then I'm not sure it should be... WDYT @sebaciv?

Reply:
`a: => A` use-case is covered by `.optIf` extension method, so this variant of `when` seems useful

Reply:
I'm just not sure this is the right API. It's misleading given the similarities to Option and we always tried to make Opt a drop-in replacement as much as possible. Maybe we should just add a method / extension method instead.

Reply:
I didn't know about the fact that `we always tried to make Opt a drop-in replacement as much as possible` If you prefer the extension method, I'd go for sth like the combination of `optIf` and `flatten`. It would be sth like a lazy filter, but with the `p: Boolean` predicate, not the `p: A: Boolean` one. So, again, it wouldn't be consistent.

Reply:
```scala
 implicit class OptOps[A](opt: => Opt[A]) {
    def lazyFilter(p: Boolean): Opt[A] = if (p) opt else Opt.Empty
  }


  Opt(5).lazyFilter(1 == 1) // Opt(5)
  Opt(5).lazyFilter(1 == 2) // Opt.Empty
  
```

Reply:
how about `.when(Boolean)` and `.unless(Boolean)` as extensions?

Reply:
It is not in line with my sense of aesthetics, but it can be ¯\_(ツ)_/¯.

Reply:
I don't mind `lazyFilter`, it's just `lazyFilterNot` that makes me a bit worried :D

Reply:
It can. You just need to avoid implicit class sugar and code it up yourself, as in `com.avsystem.commons.SharedExtensionsUtils.LazyUniversalOps`

Reply:
Imho, it is useless and misleading API.

It can be only used in the same place the Opt is created:
```scala
Opt(value).when(condition)
```
or maybe in a method, if `Opt` is passed as by-name argument or using `Function0[Opt[_]]`.
It brings nothing when it is assigned to a val, cause `.filter` can be used at this point.

`.optIf(condition).flatten` does the same and it is not much longer expression then `.when(condition)`

Besides, need of such an API, imho, means that the issue is in the implementation. Probably it should be `Opt[() => T]` or `Opt[SomeEvaluator]` or just `SomeEvaluator` that can be `optIf`ed.